### ISSUE TITLE
A possible rule to pick up `in` parameters where the caller hasn't specified the `in` keyword?

### ISSUE THREAD

**Issue by mjr4077au** [meziantou/Meziantou.Analyzer#1200]
**Is your feature request related to a problem? Please describe.**
I do a lot of Win32 development and am often passing structs by `in` reference. I've just identified a few spots where the `in` keyword wasn't in use at the call site and thought it'd be great to have an analyser rule for this.

**Describe the solution you'd like**
A rule to pick up `in` parameters where the caller hasn't specified the `in` keyword.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by meziantou** [comment]
Can you be more specific? Maybe provide some examples showing cases where `in` should be used.

**Comment by mjr4077au** [comment]
Sure thing. Consider this CsWin32 friendly overload wrapper to avoid using pointers:
```c#
internal static unsafe winmdroot.Foundation.BOOL LookupPrivilegeName([Optional] string lpSystemName, in winmdroot.Foundation.LUID lpLuid, [Optional] Span<char>lpName, ref uint cchName)
{
	fixed (uint* cchNameLocal = &cchName)
	{
		fixed (char* lpNameLocal = lpName)
		{
			fixed (winmdroot.Foundation.LUID* lpLuidLocal = &lpLuid)
			{
				fixed (char* lpSystemNameLocal = lpSystemName)
				{
					winmdroot.Foundation.BOOL __result = PInvoke.LookupPrivilegeName(lpSystemNameLocal, lpLuidLocal, lpNameLocal, cchNameLocal);
					return __result;
				}
			}
		}
	}
}
```
Ideally when `LookupPrivilegeName()` is called, I would do something like `uint bufSize; LookupPrivilegeName(null, in localLuid, null, ref bufSize)` to avoid an unnecessay copy on the LUID struct, but the `in` keyword at the call site is not enforced, therefore a copy will take place if I did `uint bufSize; LookupPrivilegeName(null, localLuid, null, ref bufSize)`.

**Comment by meziantou** [comment]
I see 2 different cases:

1. There is a single method with an `in` parameter. You want to force the call-site to add `in` when possible for readability. It doesn't change the behavior.
 
```c#
var foo = new Foo();
Sample.A(foo); // Suggest adding `in` (no behavior change)

public static class Sample
{
    public static void A(in Foo foo){}
}

public struct Foo{}
```
 
2. There are multiple overloads. Adding `in` would change the behavior to use the method with `in` parameter.

```c#
var foo = new Foo();
Sample.A(foo); // Adding `in` would select the other overload

public static class Sample
{
    public static void A(Foo foo){}
    public static void A(in Foo foo){}
}

public struct Foo{}
```

Note: `A(new Foo())` is not compatible as `A(in new Foo())` doesn't compile

In the case of the win32 generator, I think you are in case 1. So, the analyzer would not provide any benefit except clearer intent.

Am I missing something?

**Comment by mjr4077au** [comment]
It's my understanding that without the `in` keyword at the call site, if the compiler wishes to make a copy of the provided value, its free to do so, whereas the `in` keyword guards against this. It also provides a certain level of type guarding as an `in long` couldn't accept an `in int` value.

**Comment by meziantou** [comment]
I've added 2 new rules, disabled by default. You must enable them in the editorconfig file.

**Comment by mjr4077au** [comment]
Just wanted to circle back and say thanks and that it's working great.
